compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -2528,7 +2528,7 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
 
     /** Can we enumerate all instantiations of this type? */
     def isClosedSum(tp: Symbol): Boolean =
-      tp.is(Sealed) && tp.isOneOf(AbstractOrTrait) && !tp.hasAnonymousChild
+      tp.is(Sealed) && !tp.hasAnonymousChild
```

Comment:
should we check that it actually has children? and what about `sealed abstract case class`

Reply:
This method is used as a predicate to decompose types along their children (we then do `.forall` on each children). So in the case of a sealed class with zero children, the algorithm will be able correctly conclude disjointness (a sealed type without children is empty!).

I didn't think about the `sealed abstract case class` combination, but I don't think it requires any particular treatment...

Reply:
This logic relies on the contract from the `children` method:

```
/** If this is a sealed class, its known children in the order of textual occurrence */
def children(using Context): List[Symbol] =
```

If a sealed type has children that are unlisted by that method, then we are in trouble.

Reply:
I only bring up `sealed case class` because I imagine it should be considered either a sum or product with the same rule as Mirror synthesis decides, however, seeing as this is about possible sub classes at runtime then it probably does not matter

Reply:
I suspect you can't make this change without also changing the usage of `children` because now there could be values of type Foo that aren't Foo1 or Foo2.  Similar to Jamie's sealedDescendants vs sealedStrictDescendants.

Reply:
@dwijnand I see what you mean, but at the precise point where we use decompose we know that `new Foo` isn't a subtype of `tp2`, so it's safe to decompose it regardless of if it's an abstract or concrete class. To put it another way (in the context of your https://github.com/lampepfl/dotty/issues/13455#issuecomment-918251986), since at this point in the code we know that `new R` is not an instance of `Function1[X, Y]`, and because `R` is sealed, it's OK to decompose `R` using `.children`. Does this make sense?

I added a refactoring commit to try to clarify that point in the code.